internal/dcache/cluster_manager/cluster_manager.go

```diff
@@ -1289,6 +1265,51 @@ func (cmi *ClusterManager) updateRVList(existingRVMap map[string]dcache.RawVolum
 	return changed, nil
 }
 
+// For all the given NodeIds, fetch the heartbeat and return the map of RVs and map of their last heartbeat by RVId.
+func collectHBForGivenNodeIds(nodeIds []string) (map[string]dcache.RawVolume, map[string]uint64, error) {
+	rVsByRvIdFromHB := make(map[string]dcache.RawVolume)
+	rvLastHB := make(map[string]uint64)
+	for _, nodeId := range nodeIds {
+		log.Debug("ClusterManager::collectHBForGivenNodeIds: Fetching heartbeat for node %s", nodeId)
+
+		bytes, err := getHeartbeat(nodeId)
+		if err != nil {
+			common.Assert(false, err)
+			return nil, nil, fmt.Errorf("ClusterManager::collectHBForGivenNodeIds: Failed to fetch heartbeat for node %s: %v",
+				nodeId, err)
+		}
+
+		var hbData dcache.HeartbeatData
+		if err := json.Unmarshal(bytes, &hbData); err != nil {
+			common.Assert(false, err)
+			return nil, nil, fmt.Errorf("ClusterManager::collectHBForGivenNodeIds: Failed to parse heartbeat bytes (%d) for node %s: %v",
+				len(bytes), nodeId, err)
+		}
+
```

Comment:
this is the only place where we process heartbeats. Let's make sure heartbeats are valid before we process and act.
Let's add a IsValidHeartbeat() function and use it here for checking and skip the heartbeat if it's not valid. Assert for debug builds.
IsValidHeartbeat() should check that hostname is valid (not empty), ipaddress is valid, nodeid is valid and LastHeartbeat is within minUnixEpoch and maxUnixEpoch.
We can also assert than it has at least one RV (with a TODO for supporting nodes with no RVs)